### PR TITLE
Destroying Containers (chests, boxes, barrels etc) https://github.com/ot...

### PR DESCRIPTION
**Pull request by PrinterLUA**
...land/forgottenserver/issues/1014

This should be a continue of what @EvulMastah started.

### CODE REVIEW

**Comment by PrinterLUA** [comment]
I actually left that out on purpose, since i would like it to be pure lua also i feel it should be another way to do it, than make it hardcoded only because of monsters.

File: data/actions/lib/actions.lua [hunk 1]
```diff
@@ -1,5 +1,5 @@
-function destroyItem(cid, target, toPosition)
-	if target == nil or not target:isItem() then
+function destroyItem(player, target, toPosition)
+	if not target or not target:isItem() then
```
Comment (marksamman) [inline data/actions/lib/actions.lua]: What was wrong with explicitly checking for nil?

File: data/actions/lib/actions.lua [hunk 2]
```diff
@@ -7,58 +7,35 @@ function destroyItem(cid, target, toPosition)
...
-			elseif targetId == 3805 or targetId == 3806 then
-				Game.createItem(2259, 1, toPosition)
-			end
-		end
-		toPosition:sendMagicEffect(CONST_ME_POFF)
+	if toPosition.x == CONTAINER_POSITION then
+		player:sendCancelMessage(Game.getReturnMessage(RETURNVALUE_NOTPOSSIBLE))
 		return true
 	end
+	-- chest,crate, barrel...
```
Comment (marksamman) [inline data/actions/lib/actions.lua]: This comment can be removed.

**Comment by EvulMastah** [comment]
If you want it lua, you should omit Items and LuaScript changes.

File: data/actions/lib/actions.lua [hunk 3]
```diff
@@ -7,58 +7,35 @@ function destroyItem(cid, target, toPosition)
...
-		if math.random(1, 3) == 1 then
-			if targetId == 3798 then
-				target:transform(3959)
-			elseif targetId == 3799 then
-				target:transform(3958)
+		-- Move items outside the container
+		if target:isContainer() then
+			local containerItems
+			for i = target:getSize() - 1, 0, -1 do
+   				containerItems = target:getItem(i)
```
Comment (marksamman) [inline data/actions/lib/actions.lua]: Why declare containerItems local outside the loop? You can change this line to:
```lua
containerItems = target:getItem(i)
```

Comment (marksamman) [inline data/actions/lib/actions.lua]: It should also be called containerItem instead of containerItems.

File: data/actions/scripts/other/destroy.lua [hunk 4]
```diff
@@ -1,3 +1,3 @@
 function onUse(cid, item, fromPosition, target, toPosition, isHotkey)
```
Comment (marksamman) [inline data/actions/scripts/other/destroy.lua]: cid should be player

File: data/actions/scripts/tools/machete.lua [hunk 5]
```diff
@@ -4,5 +4,5 @@ function onUse(cid, item, fromPosition, target, toPosition, isHotkey)
 		target:decay()
 		return true
 	end
-	return destroyItem(cid, target, toPosition)
+	return destroyItem(player, target, toPosition)
```
Comment (marksamman) [inline data/actions/scripts/tools/machete.lua]: cid in onUse should be player

File: data/actions/scripts/tools/scythe.lua [hunk 6]
```diff
@@ -5,5 +5,5 @@ function onUse(cid, item, fromPosition, target, toPosition, isHotkey)
 		Game.createItem(2694, 1, toPosition)
 		return true
 	end
-	return destroyItem(cid, target, toPosition)
+	return destroyItem(player, target, toPosition)
```
Comment (marksamman) [inline data/actions/scripts/tools/scythe.lua]: cid in onUse should be player.

File: data/items/items.xml [hunk 7]
```diff
@@ -7195,11 +7300,9 @@
 		<attribute key="slotType" value="backpack" />
 	</item>
 	<item id="3961" article="a" name="lich staff">
-		<attribute key="description" value="A scent of death and decay sticks to this weapon." />
```
Comment (marksamman) [inline data/items/items.xml]: Why did you remove the description?

File: data/items/items.xml [hunk 8]
```diff
@@ -7195,11 +7300,9 @@
...
-		<attribute key="weight" value="4100" />
+		<attribute key="weight" value="8000" />
 		<attribute key="defense" value="30" />
 		<attribute key="attack" value="40" />
-		<attribute key="extradef" value="2" />
```
Comment (marksamman) [inline data/items/items.xml]: What's the reason to change the weight and remove extradef?

File: data/items/items.xml [hunk 9]
```diff
@@ -7257,7 +7359,7 @@
 		<attribute key="armor" value="2" />
 		<attribute key="slotType" value="head" />
 	</item>
-	<item id="3971" article="a" name="charmer's tiara">
+	<item id="3971" name="charmer's tiara" article="a">
```
Comment (marksamman) [inline data/items/items.xml]: Why did you move article?

File: data/items/items.xml [hunk 10]
```diff
@@ -7820,13 +7918,13 @@
 		<attribute key="duration" value="120" />
 	</item>
 	<item fromid="4328" toid="4341" name="" />
-	<item id="4342" name="sand">
-		<attribute key="decayTo" value="4343" />
-		<attribute key="duration" value="900" />
-	</item>
-	<item id="4343" name="sand" />
+        <item id="4342" name="sand">
```
Comment (marksamman) [inline data/items/items.xml]: You messed up the indentation here.

File: data/actions/lib/actions.lua (hunk 1 above)
Comment (PrinterLUA) [inline data/actions/lib/actions.lua]: Sorry, i'm used to my coding style ^^, should i change it back to nil?

File: data/actions/lib/actions.lua (hunk 2 above)
Comment (PrinterLUA) [inline data/actions/lib/actions.lua]: ok

File: data/actions/lib/actions.lua (hunk 3 above)
Comment (PrinterLUA) [inline data/actions/lib/actions.lua]: As i metion, my coding style. I shall change it.

**Comment by PrinterLUA** [comment]
@marksamman done, let me if there is something more i need to change.

File: data/actions/lib/actions.lua [hunk 11]
```diff
@@ -7,58 +7,33 @@ function destroyItem(cid, target, toPosition)
...
-		if math.random(1, 3) == 1 then
-			if targetId == 3798 then
-				target:transform(3959)
-			elseif targetId == 3799 then
-				target:transform(3958)
+		-- Move items outside the container
+		if target:isContainer() then
+			for i = target:getSize() - 1, 0, -1 do
+   				local containerItem = target:getItem(i)
+				if containerItems then
```
Comment (marksamman) [inline data/actions/lib/actions.lua]: containerItem

File: data/actions/lib/actions.lua [hunk 12]
```diff
@@ -7,58 +7,33 @@ function destroyItem(cid, target, toPosition)
...
+					containerItems:moveTo(toPosition)
```
Comment (marksamman) [inline data/actions/lib/actions.lua]: containerItem

File: data/items/items.xml [hunk 13]
```diff
@@ -7225,6 +7330,7 @@
 	</item>
 	<item id="3965" article="a" name="hunting spear">
 		<attribute key="weight" value="2200" />
+		<attribute key="defense" value="18" />
```
Comment (marksamman) [inline data/items/items.xml]: ?

File: data/items/items.xml [hunk 14]
```diff
@@ -7198,8 +7303,8 @@
 		<attribute key="description" value="A scent of death and decay sticks to this weapon." />
 		<attribute key="weight" value="4100" />
 		<attribute key="defense" value="30" />
-		<attribute key="attack" value="40" />
 		<attribute key="extradef" value="2" />
+		<attribute key="attack" value="40" />
```
Comment (marksamman) [inline data/items/items.xml]: Why did you move it?

File: data/items/items.xml [hunk 15]
```diff
@@ -7933,9 +8039,11 @@
 	<item id="4840" article="a" name="witches' cap spot">
 		<attribute key="weight" value="100" />
 	</item>
-	<item id="4842" article="an" name="old parchment">
+	<item id="4842" article="a" name="parchment">
```
Comment (marksamman) [inline data/items/items.xml]: ?

**Comment by PrinterLUA** [comment]
Eh, how did i miss those xD. Ok hopefully now i did not miss anything ^^

File: data/actions/lib/actions.lua [hunk 16]
```diff
@@ -7,58 +7,33 @@ function destroyItem(cid, target, toPosition)
...
-		if math.random(1, 7) == 1 then
-			if targetId == 7538 then
-				target:transform(7544)
-			elseif targetId == 7539 then
-				target:transform(7545)
-			end
-			target:decay()
+	if math.random(1) == 7 then
+		local item = Game.createItem(destroyId, 1, toPosition)
+		if item and ItemType(destroyId):getDecayId() then
```
Comment (marksamman) [inline data/actions/lib/actions.lua]: You don't need to check ItemType(destroyId):getDecayId(), just call item:decay() if item isn't nil.

File: data/actions/lib/actions.lua [hunk 17]
```diff
@@ -7,58 +7,33 @@ function destroyItem(cid, target, toPosition)
...
+   				local containerItem = target:getItem(i)
+				if containerItem then
+					containerItem:moveTo(toPosition)
+				end
 			end
-			target:decay()
 		end
+
+		target:remove(1)
 		toPosition:sendMagicEffect(CONST_ME_POFF)
```
Comment (marksamman) [inline data/actions/lib/actions.lua]: The poff effect should always appear.

File: data/actions/lib/actions.lua [hunk 18]
```diff
@@ -7,58 +7,33 @@ function destroyItem(cid, target, toPosition)
...
-	--spider web
-	if targetId == 7538 or targetId == 7539 then 
-		if math.random(1, 7) == 1 then
-			if targetId == 7538 then
-				target:transform(7544)
-			elseif targetId == 7539 then
-				target:transform(7545)
-			end
-			target:decay()
+	if math.random(1) == 7 then
```
Comment (marksamman) [inline data/actions/lib/actions.lua]: Please enlighten me how this can ever be true.

Comment (EPuncker) [inline data/actions/lib/actions.lua]: lol

**Comment by PrinterLUA** [comment]
:facepalm:

**Comment by marksamman** [comment]
Test your changes to confirm that it actually works, since I suspect that this hasn't been tested, and then squash your commits.

File: data/actions/lib/actions.lua [hunk 19]
```diff
@@ -7,58 +7,34 @@ function destroyItem(cid, target, toPosition)
...
-			elseif targetId == 3805 or targetId == 3806 then
-				Game.createItem(2259, 1, toPosition)
-			end
-		end
-		toPosition:sendMagicEffect(CONST_ME_POFF)
+	if toPosition.x == CONTAINER_POSITION then
+		player:sendCancelMessage(Game.getReturnMessage(RETURNVALUE_NOTPOSSIBLE))
 		return true
 	end
+	local destroyId = ItemType(target.itemid):getDestroyId()
```
Comment (ninjalulz) [inline data/actions/lib/actions.lua]: Add a new line above this variable.

File: data/items/items.xml [hunk 20]
```diff
@@ -6851,8 +6937,13 @@
 	<item id="3792" article="a" name="stone pillar" />
 	<item id="3793" article="a" name="broken stone pillar" />
 	<item fromid="3794" toid="3797" name="wooden bars" />
-	<item fromid="3798" toid="3799" name="wooden bars">
+	<item id="3798" name="wooden bars">
+		<attribute key="description" value="They already have some cracks and look rather fragile." />
+		<attribute key="destroyTo" value="3959" />
+	</item>
+	<item id="3799" name="wooden bars" >
```
Comment (ninjalulz) [inline data/items/items.xml]: Remove space before the greater-than sign.

File: data/items/items.xml [hunk 21]
```diff
@@ -8002,7 +8107,7 @@
 		<attribute key="fluidSource" value="blood" />
 	</item>
 	<item id="4860" name="wooden trash" />
-	<item id="4861" article="a" name="cobra statue"/>
+	<item id="4861" article="a" name="cobra statue" />
 	<item id="4862" name="destroyed cobra statue" >
```
Comment (ninjalulz) [inline data/items/items.xml]: Remove space before the greater-than sign.

File: data/items/items.xml [hunk 22]
```diff
@@ -13055,7 +13161,13 @@
 	</item>
 	<item id="7536" name="the remains of a spider egg" />
 	<item id="7537" article="a" name="spider egg" />
-	<item fromid="7538" toid="7543" article="a" name="spider web" />
+	<item id="7538" article="a" name="spider web" >
```
Comment (ninjalulz) [inline data/items/items.xml]: Remove space before the greater-than sign.

**Comment by PrinterLUA** [comment]
Everything seems to work:
http://1.ii.gl/MCLtpFj0x.gif

**Comment by marksamman** [comment]
Apply the changes from @ninjalulz comments, and when you squash your commits again, please write a good commit message.

**Comment by PrinterLUA** [comment]
Ok, what is good squash commit message? Is it the message which appears, when you merge it?

**Comment by marksamman** [comment]
"Update destroyItem to use new destroyTo item property"

**Comment by PrinterLUA** [comment]
Done :)

**Comment by marksamman** [comment]
Thanks!

**Comment by PrinterLUA** [comment]
Np :)
